apps/e2e/tests/backend/endpoints/api/v1/internal/failed-emails-digest.test.ts

```diff
@@ -180,4 +158,240 @@ describe("with valid credentials", () => {
     const digestEmail = messages.find(msg => msg.subject === "Failed emails digest");
     expect(digestEmail).toBeUndefined();
   });
+
+  it("should not send digest email when project owner has no primary email", async ({ expect }) => {
+    backendContext.set({
+      projectKeys: InternalProjectKeys,
+      userAuth: null,
+    });
+    const { userId } = await Auth.Otp.signIn();
+    const adminAccessToken = backendContext.value.userAuth?.accessToken;
+    expect(adminAccessToken).toBeDefined();
+    await Project.create({
+      display_name: "Test Project No Owner Email",
+    });
+
+    // Remove primary email from the user
+    const updateEmailResponse = await niceBackendFetch(`/api/v1/users/${userId}`, {
+      method: "PATCH",
+      accessType: "admin",
+      body: {
+        "primary_email": null,
+      },
+    });
+    expect(updateEmailResponse.status).toBe(200);
+
+    // Send a test email that will fail
+    await niceBackendFetch("/api/v1/internal/send-test-email", {
+      method: "POST",
+      accessType: "admin",
+      headers: {
+        "x-stack-admin-access-token": adminAccessToken,
+      },
+      body: {
+        "recipient_email": "test-email-recipient@stackframe.co",
+        "email_config": {
+          "host": "this-is-not-a-valid-host.example.com",
+          "port": 123,
+          "username": "123",
+          "password": "123",
+          "sender_email": "123@g.co",
+          "sender_name": "123"
+        }
+      },
+    });
+
+    const initialMessageCount = (await backendContext.value.mailbox.fetchMessages()).length;
+    const response = await niceBackendFetch("/api/v1/internal/failed-emails-digest", {
+      method: "POST",
+      headers: { "Authorization": "Bearer mock_cron_secret" }
+    });
+    expect(response.status).toBe(200);
+
+    const messages = await backendContext.value.mailbox.fetchMessages();
+    const newMessages = messages.slice(initialMessageCount);
```

Comment:
The tests will run in parallel, so this won't work. Instead, you'll have to filter by email (as you did in other tests)

Reply:
this also applies to everywhere else where you use .slice